File: koguchi/2/solution_part1.py
from typing import List


def parse_lines(lines: List) -> List:
    reports = []
    for line in lines:
        line = line.strip()
        report = line.split(' ')
        reports.append([int(level) for level in report])

    return reports


with open('input.txt', 'r') as f:
    raw_input = f.readlines()
    reports = parse_lines(raw_input)

# print(reports)


def is_within_safe_range(current: int, previous: int) -> bool:
    within_safe_range = 1 <= abs(current - previous) <= 3
    # if not within_safe_range:
    #     print(f'difference: {abs(current - previous)}')
    return within_safe_range


def changes_direction(current: int, previous: int, is_descending: bool) -> bool:
    return (current > previous) == is_descending


def report_is_safe(report):
    is_descending = None
    for i in range(1, len(report)):
        previous = report[i-1]
        current = report[i]
        # print(previous, current)

        if not is_within_safe_range(current, previous):
            # print('Breaking unsafe range')
            break

        if is_descending is None:
            is_descending = current < previous
            # print(f'This run is descending: {is_descending}')

        if changes_direction(current, previous, is_descending):
            # print('Breaking change direction')
            break
    else:
        return True

    return False


# O(num_reports*levels)
num_safe = 0
for report in reports:
    print(report)

    if report_is_safe(report):
        num_safe += 1

print(num_safe)

File: koguchi/02/solution_part1.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>
#include <sstream>


using namespace std;

const int NUM_REPORTS = 1000;
const int UPPER_BOUND = 3;
const int LOWER_BOUND = 1;


int main() {

    vector<vector<int>> reports;

    string line;
    while (getline(cin, line)) {
        vector<int> report;
        istringstream iss(line);

        int level;
        while (iss >> level) {
            report.push_back(level);
        }
        reports.push_back(report);
    }

    int numSafe = NUM_REPORTS;
    for (auto report: reports) {
        bool isDescending = report[0] > report[1];

        for (int i=1; i<report.size(); i++) {
            cout << report[i-1] << ' ' << report[i] << " - ";

            // Changes direction
            if (isDescending) {
                if (report[i] > report[i-1]) {
                    numSafe--;
                    break;
                }
            }
            else {
                if (report[i] < report[i-1]) {
                    numSafe--;
                    break;
                }
            }

            // Safe levels
            int levelDiff = abs(report[i] - report[i-1]);
            if (levelDiff > UPPER_BOUND || levelDiff < LOWER_BOUND) {
                numSafe--;
                break;
            }
        }

        cout << "Num Safe: " << numSafe << endl;
    }

    return 0;
}

